src/download-project-actions.ts

```diff
@@ -0,0 +1,202 @@
+import {
+  MicrobitWebUSBConnection,
+  ConnectionStatus as UsbConnectionStatus,
+} from "@microbit/microbit-connection";
+import {
+  ConnectActions,
+  ConnectAndFlashResult,
+  ConnectionAndFlashOptions,
+} from "./connect-actions";
+import { ConnectionStatus } from "./connect-status-hooks";
+import { ConnectionStageActions } from "./connection-stage-actions";
+import {
+  DownloadProjectStage,
+  DownloadProjectStep,
+  MicrobitToFlash,
+} from "./download-project-hooks";
+
+export class DownloadProjectActions {
+  constructor(
+    private stage: DownloadProjectStage,
+    private setStage: (stage: DownloadProjectStage) => void,
+    private connectActions: ConnectActions,
+    private connectionStageActions: ConnectionStageActions,
+    private setFlashingProgress: (progress: number) => void,
+    private connectionStatus: ConnectionStatus
+  ) {}
+
+  clearMakeCodeUsbDevice = () => {
+    this.setStage({ ...this.stage, usbDevice: undefined });
+  };
+
+  start = async (download: { name: string; hex: string }) => {
+    if (this.stage.usbDevice) {
+      if (this.stage.usbDevice.status === UsbConnectionStatus.CONNECTED) {
+        const newStage: DownloadProjectStage = {
+          ...this.stage,
+          step: DownloadProjectStep.FlashingInProgress,
+          projectHex: download.hex,
+          projectName: download.name,
+        };
+        this.setStage(newStage);
+        return this.flashMicrobit(newStage, {
+          temporaryUsbConnection: this.stage.usbDevice,
+        });
+      }
+      return this.updateStage({
+        projectHex: download.hex,
+        projectName: download.name,
+        step: DownloadProjectStep.ConnectCable,
+      });
+    }
+    const projectInfo = {
+      projectHex: download.hex,
+      projectName: download.name,
+    };
+    if (this.stage.skipIntro) {
+      return this.onIntroNext(projectInfo);
+    }
+    this.updateStage({
+      step: DownloadProjectStep.Introduction,
+      ...projectInfo,
+    });
+  };
+
+  onIntroNext = async (partialNewStage: Partial<DownloadProjectStage> = {}) => {
+    if (this.connectionStatus === ConnectionStatus.Connected) {
+      return this.updateStage({
+        ...partialNewStage,
+        step: DownloadProjectStep.ChooseSameOrAnotherMicrobit,
+      });
+    }
+    // TODO: Is the right behaviour?
+    // Maybe the user should choose a device if they have previously
+    // connected to one, even if it isn't connected right now.
```

Comment:
Flagging todo - this PR.

Reply:
We agreed that if a device was previously connected, we shouldn't automatically overwrite it, even if it's not necessary connected at the time of MakeCode hex flashing [cd96c31](https://github.com/microbit-foundation/ml-trainer/pull/301/commits/cd96c317abd2d9abb199791250bbc6c60a013be5)